controllers/singleton_installer.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"sync/atomic"
+)
+
+type InstallStatus int32
+
+const (
+	InstallStatusUnknown InstallStatus = iota
+	InstallStatusInstalling
+	InstallStatusReady
+)
+
+func (i InstallStatus) String() string {
+	switch i {
+	case InstallStatusInstalling:
+		return "Installing"
+	case InstallStatusReady:
+		return "Ready"
+	default:
+		return "Unknown"
+	}
+}
+
+type singletonInstaller struct {
+	status *InstallStatus
+}
+
+type InstallFn func() error
+
+// SingletonInstaller is a simple interface to ensure that only one thread
+// can run an install at a time.
+type SingletonInstaller interface {
+	Install(InstallFn) (InstallStatus, error)
+}
+
+func NewSingletonInstaller() SingletonInstaller {
+	si := &singletonInstaller{status: new(InstallStatus)}
+	atomic.StoreInt32((*int32)(si.status), int32(InstallStatusUnknown))
+	return si
+}
+
+func (s *singletonInstaller) getStatus() InstallStatus {
+	return InstallStatus(atomic.LoadInt32((*int32)(s.status)))
+}
+
+// compareAndSwap will return false if oldStatus is not as expected.
+func (s *singletonInstaller) setStatus(oldStatus, newStatus InstallStatus) bool {
+	return atomic.CompareAndSwapInt32((*int32)(s.status), int32(oldStatus), int32(newStatus))
+}
+
+func (s *singletonInstaller) Install(installFn InstallFn) (InstallStatus, error) {
+	for {
+		status := s.getStatus()
+		if status == InstallStatusReady || status == InstallStatusInstalling {
+			// nothing to do as we are in either of:
+			// 1. Ready: already installed
+			// 2. Installing: another thread is busy installing
+			return status, nil
+		}
+		// we must definitally own the "installing" status before beginning the Install.
+		// if another thread has caused this to be in a different status then we need
+		// to go around the loop again.
+		if s.setStatus(InstallStatusUnknown, InstallStatusInstalling) {
+			break
+		}
+	}
+
+	err := installFn()
```

Comment:
If there's an error here, how will I know? We don't log it or set it on any status